File: src/antpack/numbering_tools/annotator_base_class.py
"""This is the base class for Single and MultiChainAnnotator. Contains routines
shared by both."""
from antpack_cpp_ext import sort_position_codes_cpp




class AnnotatorBaseClass:

    def __init__(self, scheme):
        if scheme not in ["imgt", "martin", "kabat"]:
            raise ValueError("Unsupported scheme supplied.")

        self.scheme = scheme
        self.chain_maps = {"H":"H", "K":"L", "L":"L"}


    def sort_position_codes(self, position_code_list):
        """Convenience function that takes as input a list of position codes
        in any order and sorts them in a way that respects the properties of
        the numbering scheme. This is not as trivial as it appears, since
        in the IMGT scheme for example '111', '111A', '112B', '112A', '112" is
        correct ordering (but '86A', '87B', '87A' would not be). This is useful
        if you have a list of all of the unique position codes observed in a large
        dataset and would like to sort it so you can then encode all of these
        sequences as fixed-length arrays.

        Note that '-' is not a valid code so it is removed if it is present in the
        list passed to this function (you do not have to remove it yourself).

        Args:
            position_code_list (list): A list of strings. Each must be a valid position
                code for the scheme selected for this Annotator.

        Returns:
            sorted_code_list (list): A list of sorted position codes.

        Raises:
            RuntimeError: A runtime error is raised if invalid codes are supplied.
        """
        # Passing '-' to this function is a common mistake, so we check for it first.
        cleaned_codes = [c for c in position_code_list if c != "-"]
        sorted_codes, is_valid = sort_position_codes_cpp(cleaned_codes, self.scheme)
        if not is_valid:
            raise RuntimeError("One or more of the position codes supplied was rejected as "
                    "invalid by the sorter.")
        return sorted_codes


    def build_msa(self, sequences, annotations):
        """Builds a multiple sequence alignment using a list of sequences and a
        list of tuples output by analyze_seq or analyze_seqs (PairedChainAnnotator
        or SingleChainAnnotator).

        Args:
            sequences (list): A list of sequences.
            annotations (list): A list of tuples, each containing (numbering,
                percent_identity, chain_name, error_message). These tuples
                are what you will get as output if you pass sequences to
                the analyze_seq or analyze_seqs methods of SingleChainAnnotator
                or PairedChainAnnotator.

        Returns:
            position_codes (list): A list of position codes from the appropriate numbering
                scheme.
            aligned_seqs (list): A list of strings -- the input sequences all aligned
                to form an MSA.

        Raises:
            RuntimeError: A RuntimeError is raised if 1) sequences of different chaintypes
                are passed, 2) incorrect input is passed.
        """
        if not isinstance(annotations, list) or not isinstance(sequences, list):
            raise RuntimeError("sequences and annotations should both be lists.")

        if len(sequences) != len(annotations):
            raise RuntimeError("The length of sequences and annotations did not match.")

        position_codes = set()
        chain_types = set()

        for annotation in annotations:
            if len(annotation) != 4:
                raise RuntimeError("Each annotation should be a tuple with four elements: "
                        "numbering, percent identity, chain name, error message. "
                        "In general it is easiest to pass the output of analyze_seq() to this "
                        "function.")
            if not isinstance(annotation[0], list) or not isinstance(annotation[1], float) \
                or not isinstance(annotation[2], str) or not isinstance(annotation[3], str):
                raise RuntimeError("Each annotation should be a tuple with four elements: "
                        "numbering, percent identity, chain name, error message. "
                        "In general it is easiest to pass the output of analyze_seq() to this "
                        "function.")
            for position_code in annotation[0]:
                if position_code != "-":
                    position_codes.add(position_code)

            chain_types.add(self.chain_maps[annotation[2]])

        if len(chain_types) > 1:
            raise RuntimeError("You tried to build an MSA out of two different chain types!")

        position_codes = self.sort_position_codes(list(position_codes))

        position_dict = {k:i for i, k in enumerate(position_codes)}
        msa = [['-' for _ in position_codes] for s in sequences]

        for i, (sequence, annotation) in enumerate(zip(sequences, annotations)):
            for position, letter in zip(annotation[0], sequence):
                if position == "-":
                    continue
                msa[i][position_dict[position]] = letter

        msa = ["".join(m) for m in msa]
        return position_codes, msa


    def trim_alignment(self, sequence, alignment):
        """Takes as input a sequence and atuple produced by
        analyze_seq and trims off any gap regions at the end
        that result when there are amino acids on either end
        which are not part of the numbered variable region.
        The output from analyze_seq can be fed directly to this
        function.

        Args:
            sequence (str): The sequence corresponding to the numbering / alignment.
            alignment (tuple): A tuple of any length where the first element is
                the sequence_numbering list returned by analyze_seq.
                sequence_numbering must be of the same length as sequence.

        Returns:
            trimmed_sequence (str): The input sequence trimmed so that non-numbered AAs on
                either end have been removed.
            trimmed_numbering (list): The sequence_numbering from the input trimmed so that
                any '-' assignments on either end have been removed.
            exstart (int): The start of the extracted region represented by trimmed_sequence.
            exend (int): The end of the extracted region represented by trimmed_sequence.
                trimmed_sequence in other words is sequence[start:end].

        Raises:
            RuntimeError: A RuntimeError is raised if unexpected inputs are supplied.
        """
        if not isinstance(alignment, tuple):
            raise RuntimeError("Argument to trim_alignment must be a tuple.")

        numbering = alignment[0]
        if len(numbering) != len(sequence):
            raise RuntimeError("Sequence and numbering should be the same length to "
                    "use this function.")
        exstart = next((i for i in range(len(numbering)) if numbering[i] != '-'), 0)
        exend = next((i for i in range(exstart + 1, len(numbering)) if numbering[i] == '-'),
                len(numbering))
        trimmed_sequence = sequence[exstart:exend]
        trimmed_numbering = numbering[exstart:exend]

        return trimmed_sequence, trimmed_numbering, exstart, exend

File: src/antpack/cpp_src/annotator_classes/annotator_base_class.cpp
/* Copyright (C) 2025 Jonathan Parkinson
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
// C++ headers
#include <tuple>
#include <vector>
#include <string>
#include <memory>
#include <unordered_map>

// Project headers
#include "annotator_base_class.h"
#include "../utilities/cdr_assignment_utilities.h"

namespace SequenceAnnotators {


AnnotatorBaseClassCpp::AnnotatorBaseClassCpp(std::string scheme):
scheme(scheme) {
}





/// AnnotatorBaseClassCpp function which sorts a list of
/// position codes (essentially by calling the corresponding
/// function in utilities). Essentially a wrapper on the
/// corresponding utility function that can be accessed by
/// Python or outside callers.
std::vector<std::string> AnnotatorBaseClassCpp::sort_position_codes(
        std::vector<std::string> position_code_list) {
    std::vector<std::string> cleanedCodes, sortedCodes;
    int isValid;

    for (auto &code : position_code_list) {
        if (code != "-")
            cleanedCodes.push_back(code);
    }
    if (!SequenceUtilities::sort_position_codes_utility(cleanedCodes,
                this->scheme, sortedCodes)) {
        throw std::runtime_error(std::string("One or more of the "
                    "supplied position codes is invalid given the "
                    "specified scheme."));
    }
    return sortedCodes;
}



/// AnnotatorBaseClassCpp function which builds a multiple sequence alignment from
/// a list of previously numbered sequences. The sequences must all be of the
/// same chain type. Essentially a wrapper on the corresponding utility function
/// that can be accessed by Python or outside callers.
std::tuple<std::vector<std::string>, std::vector<std::string>>
AnnotatorBaseClassCpp::build_msa(std::vector<std::string> sequences,
        std::vector<std::tuple<std::vector<std::string>, double,
        std::string, std::string>> annotations,
        bool add_unobserved_positions) {
    std::vector<std::string> positionCodes, alignedSeqs;

    if (!SequenceUtilities::build_msa_utility(sequences, annotations, positionCodes,
            alignedSeqs, this->scheme, add_unobserved_positions)) {
        throw std::runtime_error(std::string("A fatal error occured when "
                    "building an MSA -- please report."));
    }

    return std::tuple<std::vector<std::string>,
           std::vector<std::string>>{positionCodes, alignedSeqs};
}


/// AnnotatorBaseClassCpp function which trims an alignment to
/// remove gap positions at either end. Essentially a wrapper
/// on the corresponding utility function that can
/// be accessed by Python or outside callers.
std::tuple<std::string, std::vector<std::string>, int, int>
    AnnotatorBaseClassCpp::trim_alignment(std::string sequence,
        std::tuple<std::vector<std::string>, double,
        std::string, std::string> alignment) {
    int exstart, exend;
    std::vector<char> trimmedSeq;
    std::vector<std::string> trimmedAlignment;

    if (!SequenceUtilities::trim_alignment_utility(sequence,
                alignment, trimmedAlignment, exstart, exend,
                trimmedSeq)) {
        throw std::runtime_error(std::string("Invalid sequence / "
                    "alignment pairing supplied."));
    }

    std::string trimmedSeqStr(trimmedSeq.begin(), trimmedSeq.end());

    return std::tuple<std::string, std::vector<std::string>,
           int, int>{trimmedSeqStr, trimmedAlignment, exstart, exend};
}


/// Wraps the assign_cdr_labels function in SequenceUtilities
/// for access by Python callers.
std::vector<std::string> AnnotatorBaseClassCpp::assign_cdr_labels(
        std::vector<std::string> numbering, std::string chain,
        std::string scheme) {
    std::vector<std::string> cdr_labeling;
    if (scheme == "") {
        CDRConversionUtilities::assign_cdr_labels(numbering, chain,
            cdr_labeling, this->scheme, this->scheme);
    } else {
        CDRConversionUtilities::assign_cdr_labels(numbering, chain,
            cdr_labeling, this->scheme, scheme);
    }
    return cdr_labeling;
}

}  // namespace SequenceAnnotators
